runtime/stdlib/cadence_v0.42_to_v1_contract_upgrade_validator.go

```diff
@@ -393,37 +450,91 @@ typeSwitch:
 					newTypeArg := instantiationType.TypeArguments[0]
 
-					return validator.checkTypeUpgradability(oldTypeArg.Type, newTypeArg.Type)
+					return validator.checkTypeUpgradability(oldTypeArg.Type, newTypeArg.Type, true)
 				}
 			}
 		}
 
 	case *ast.NominalType:
-		if validator.checkUserDefinedType == nil {
-			break
-		}
 
-		if _, isbuiltinType := builtinTypes[oldType.String()]; !isbuiltinType {
-			checked, valid := validator.checkUserDefinedTypeCustomRules(oldType, newType)
+		oldTypeName := oldType.String()
 
-			// If there are no custom rules for this type,
-			// do the default type comparison.
-			if !checked {
-				break
+		if validator.checkUserDefinedType != nil {
+			if _, isbuiltinType := builtinTypes[oldTypeName]; !isbuiltinType {
+				checked, valid := validator.checkUserDefinedTypeCustomRules(oldType, newType)
+
+				// If there are no custom rules for this type,
+				// do the default type comparison.
+				if !checked {
+					break
+				}
+
+				if valid {
+					return nil
+				}
+
+				return newTypeMismatchError(oldType, newType)
 			}
+		}
 
-			if valid {
-				return nil
+		var expectedTypeName string
+
+		isAccountType := true
+
+		switch oldTypeName {
+		case "AuthAccount":
+			return newType.CheckEqual(astFullyEntitledAccountReference, validator)
+
+		case "PublicAccount":
+			return newType.CheckEqual(astAccountReference, validator)
+
+		case "AuthAccount.Capabilities",
+			"PublicAccount.Capabilities":
+			expectedTypeName = sema.Account_CapabilitiesType.QualifiedString()
+
+		case "AuthAccount.AccountCapabilities":
+			expectedTypeName = sema.Account_AccountCapabilitiesType.QualifiedString()
+
+		case "AuthAccount.StorageCapabilities":
+			expectedTypeName = sema.Account_StorageCapabilitiesType.QualifiedString()
+
+		case "AuthAccount.Contracts",
+			"PublicAccount.Contracts":
+			expectedTypeName = sema.Account_ContractsType.QualifiedString()
+
+		case "AuthAccount.Keys",
+			"PublicAccount.Keys":
+			expectedTypeName = sema.Account_KeysType.QualifiedString()
+
+		case "AuthAccount.Inbox":
+			expectedTypeName = sema.Account_InboxType.QualifiedString()
+
+		case "AccountKey":
+			expectedTypeName = sema.AccountKeyType.QualifiedString()
+
+		default:
+			isAccountType = false
+
+		}
+
+		if isAccountType {
+			// Only reaches here for the deprecated account types.
+			if newType.String() != expectedTypeName {
+				return newTypeMismatchError(oldType, newType)
 			}
+			return nil
 
-			return newTypeMismatchError(oldType, newType)
 		}
 	}
 
 	// If the new/old type is non-storable,
 	// then changing the type of this field has no impact to the storage.
-	if isNonStorableType(oldType) || isNonStorableType(newType) {
-		return nil
+	// However, fields having non-storable types inside capabilities are in-fact storable.
+	// So, skip only if the non-storable type is a direct field type.
+	if !inCapability {
+		if isNonStorableType(oldType) || isNonStorableType(newType) {
+			return nil
+		}
```

Comment:
This is a separate issue that I discovered while fixing the original issue: it is only OK to change to/from a non-storable type (hence the field itself is also non-storable), only if the type was a direct field-type. i.e: if the non-storabe type was residing inside a Capability, then the field is in-fact storable, and changing the type is NOT OK.